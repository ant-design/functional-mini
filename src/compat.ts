export * from './3rd-party/compat.js';

import compat from './3rd-party/compat.js';
export default compat;
