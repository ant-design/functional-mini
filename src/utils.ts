let logSwitch = true;

export const setLogSwitch = (on: boolean) => {
  logSwitch = on;
};

//@ts-expect-error
export const log = (...args) => {
  if (!logSwitch) return;

  console.log.apply(console, args);
};

//@ts-expect-error
export const error = (...args) => {
  console.error.apply(console, args);
};

export const time = (label?: string) => {
  if (!logSwitch) return;

  console.time(label);
};

export const timeEnd = (label?: string) => {
  if (!logSwitch) return;
  console.timeEnd(label);
};

//@ts-expect-error
export const shallowCompare = (obj1, obj2) => {
  if (obj1 === obj2) return true;
  if (typeof obj1 !== typeof obj2) return false;
  if (typeof obj1 !== 'object' || typeof obj2 !== 'object') return false;
  if (obj1 === null || obj2 === null) return false;
  if (Array.isArray(obj1) || Array.isArray(obj2)) return false;
  if (Object.keys(obj1).length !== Object.keys(obj2).length) return false;
  for (const key in obj1) {
    if (obj1[key] !== obj2[key]) return false;
  }
  return true;
};

// 这么做是为了测试方便，隔离一些全局变量
export enum EComponent2Status {
  VALID = 'VALID',
  INVALID = 'INVALID',
  UNKNOWN = 'UNKNOWN',
}

let component2Status = EComponent2Status.UNKNOWN;
export const getComponent2Status = () => {
  return component2Status;
};

export const updateComponent2Status = (status?: EComponent2Status) => {
  if (status) {
    component2Status = status;
    //@ts-expect-error
  } else if (typeof my !== 'undefined') {
    //@ts-expect-error
    component2Status = my.canIUse('component2')
      ? EComponent2Status.VALID
      : EComponent2Status.INVALID;
  }
};

let incrementalId = 0;
const instanceKeyId = '_functional_instance_id_';

//@ts-expect-error
export function getIdFromAppxInstance(appxInstance) {
  if (!appxInstance) throw new Error('appxInstance param is falsy');
  let id;
  // 支付宝端
  if (appxInstance.$id) {
    id = `${appxInstance.$id}-${appxInstance.$page?.$viewId || ''}-${
      appxInstance.is || 'page'
    }`; /* 页面 */
  } else if (appxInstance[instanceKeyId]) {
    id = appxInstance[instanceKeyId];
  } else {
    id = `instance-${incrementalId++}`;
    appxInstance[instanceKeyId] = id;
  }
  if (!id) throw new Error('failed to identify appxInstance.$id');
  return id;
}

export const instanceKeyPropNames = '_functional_instance_prop_names_';

export function mergeObjectKeys(
  ...objects: Record<string, unknown>[]
): string[] {
  let res: string[] = [];
  for (const iterator of objects) {
    res = res.concat(Object.keys(iterator));
  }
  return res;
}
