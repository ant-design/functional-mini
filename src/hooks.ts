import { useEffect, createContext, useContext } from './r.js';
import HandlersController from './handlers.js';
import { instanceKeyPropNames } from './utils.js';
import {
  IPlatformConstants,
  checkIfPlatformIsLoadCorrectly,
} from './platform.js';
import { ETargetPlatform } from './types.js';

export type THooksFn = (handler: Function, deps?: any[]) => void;

//@ts-expect-error
function generateEventHookName(eventName) {
  return `use${eventName[0].toUpperCase()}${eventName.slice(1)}`;
}

export const reactContext = createContext<IElementContext | null>(null);

export interface IElementContext {
  instance: any;
  handlersController: HandlersController;
  ifServerRender?: boolean;
  debugLog?: (...args: any[]) => void;
  platformConfig: IPlatformConstants;
}

// --------------------------
function useAppxContext(): IElementContext {
  const appxInstanceContext = useContext<IElementContext | null>(reactContext);
  if (!appxInstanceContext) {
    throw new Error('请不要在组件内调用 hooks');
  }
  return appxInstanceContext;
}

function useEventCall(
  name: string,
  handler: Function,
  deps: any[],
  disableMultiImpl: boolean,
) {
  if (!deps)
    console.warn(
      `useEventCall ${name}: hooks 的 deps 参数为空，可能会导致性能问题`,
    );

  const appxInstanceContext = useAppxContext();
  if (appxInstanceContext.ifServerRender) {
    // 虚拟渲染时，注册空实现
    appxInstanceContext.handlersController.addHandler(
      name,
      {},
      () => {},
      false,
    );
  }

  useEffect(() => {
    const off = appxInstanceContext.handlersController.addHandler(
      name,
      appxInstanceContext.instance,
      //@ts-expect-error
      function (this, ...args) {
        return handler.apply(undefined, args);
      },
      disableMultiImpl,
    );

    return off;
  }, deps);
}

// 注册和更新 handler，注意只能更新第一次注册过的 handler 实现，不允许变更数量和 key
export function useEvent(name: string, handler: Function, deps: any[]) {
  const appxInstanceContext = useAppxContext();
  const { platformConfig } = appxInstanceContext;
  const { pageEvents, componentEvents, blockedProperty } = platformConfig;
  if (pageEvents.indexOf(name) >= 0 || componentEvents.indexOf(name) >= 0) {
    throw new Error(
      `小程序 ${name} 是生命周期相关的保留方法，不允许使用此种方法注册，请使用对应的 hooks: ${generateEventHookName(
        name,
      )}`,
    );
  } else if (blockedProperty.indexOf(name) >= 0) {
    throw new Error(
      `不允许注册名为 ${name} 的事件处理函数，这是小程序的保留属性，请换一个名称`,
    );
  }

  useEventCall(name, handler, deps, true);
}

export function getLifeCycleHooks(
  eventName: string,
  disableMultiImpl = false,
  specifyPlatform?: ETargetPlatform,
): THooksFn {
  return (handler, deps?: any[]) => {
    const appxInstanceContext = useAppxContext();
    if (specifyPlatform) {
      const { platformConfig } = appxInstanceContext;
      checkIfPlatformIsLoadCorrectly(platformConfig, specifyPlatform);
    }

    useEventCall(eventName, handler, deps ?? [], disableMultiImpl);
  };
}

export function useSyncMiniData(data = {}) {
  const appxInstanceContext = useAppxContext();
  // const propKeys = appxInstanceContext
  if (!appxInstanceContext.instance) {
    throw new Error('cannot get appx instance, failed to set data');
  }
  if (appxInstanceContext.ifServerRender) {
    appxInstanceContext.instance.setData(data);
  }

  const debugLog = appxInstanceContext.debugLog || function () {};
  if (typeof data !== 'object')
    throw new Error(`函数返回的数据必须是一个对象，收到了 ${typeof data}`);
  // 这里是一个每次都要跑的 passive effect
  useEffect(() => {
    const { instance } = appxInstanceContext;
    const propNames = instance[instanceKeyPropNames] || []; // 微信的 data 里包含了 props，要手动踢掉
    const pendingData = {};
    const previousData = instance.data || {};

    // 比对一下，只 set 不同的部分
    for (const key in data) {
      if (!Object.prototype.hasOwnProperty.call(data, key)) continue;
      //@ts-expect-error
      if (typeof data[key] === 'function')
        throw new Error(`${key} - 不允许传入函数类型的数据`); // 暂不支持，有需求再说

      //@ts-expect-error
      if (!previousData[key] || previousData[key] !== data[key]) {
        //@ts-expect-error
        pendingData[key] = data[key];
      }
    }

    // 缺少某些 key，就设置成 null
    for (const key of Object.keys(previousData)) {
      if (propNames.indexOf(key) >= 0) continue;
      if (!Object.prototype.hasOwnProperty.call(data, key)) {
        //@ts-expect-error
        pendingData[key] = null;
      }
    }

    if (Object.keys(pendingData).length > 0) {
      debugLog('calling setData', pendingData);
      instance.setData(pendingData);
    }
  }, undefined as any);
}

export function useWechatTriggerEvent() {
  const appxInstanceContext = useAppxContext();
  const { platformConfig } = appxInstanceContext;
  checkIfPlatformIsLoadCorrectly(platformConfig, ETargetPlatform.wechat);
  const ifSSr = appxInstanceContext.ifServerRender;
  const fn = appxInstanceContext.instance?.triggerEvent;
  if (!ifSSr && !fn) {
    throw new Error('cannot get triggerEvent function from instance');
  }
  return fn;
}
