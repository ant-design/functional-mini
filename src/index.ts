import { setLogSwitch } from './utils.js';

export { setLogSwitch };
