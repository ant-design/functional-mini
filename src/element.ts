import { React, VNode, act, mountElement, serverRender } from './r.js';
import {
  log as globalLog,
  error as globalError,
  EComponent2Status,
  getComponent2Status,
  updateComponent2Status,
  getIdFromAppxInstance,
  instanceKeyPropNames,
} from './utils.js';

import HandlersController from './handlers.js';
import { IElementContext, reactContext, useSyncMiniData } from './hooks.js';
import { ETargetPlatform, EElementType } from './types.js';
import { platformConfig } from './platform.js';

const FUNCTIONAL_MINI_PAGE_DOM_PLACEHOLDER = 'MINIFISH_PAGE_DOM_PLACEHOLDER';
const DANGER_ZONE_BYPASS_FUNCTION_CALL_WITH_DATA =
  'DANGER_ZONE_BYPASS_FUNCTION_CALL_WITH_DATA'; // 绕过所有函数方法，直接返回 data。不要在生产环境使用！

export interface IAppxOptions {
  data: Record<string, any>;
  options: Record<string, string>;
  methods?: Record<string, (args?) => void>;
  [handlerName: string]: any;
}

export type TElementFunction<TProps> = (props: TProps) => {
  [axmlKey: string]: any;
};

export interface IInstanceMap {
  [id: string]: {
    appxContext: IElementContext;
    appxId: string;
    elementFn: any;
    elementInstance?: any;
    pendingProps: any;
    unmounted: boolean;
    propKeys: string[];
  };
}

function compositeElementWithContext(
  id,
  elementFn,
  appxContext: IElementContext,
  pendingProps,
) {
  const el = React.createElement(elementFn, pendingProps);
  // eslint-disable-next-line react/no-children-prop
  return React.createElement(
    reactContext.Provider,
    { value: appxContext, key: id, children: [el] },
    el,
  ) as VNode;
}

export function flushReactTree(elementMap: IInstanceMap) {
  const children: VNode[] = [];
  for (const id in elementMap) {
    if (!Object.prototype.hasOwnProperty.call(elementMap, id)) continue;

    const item = elementMap[id];
    if (item.unmounted) {
      continue;
    }
    if (item.elementInstance && !item.pendingProps) {
      children.push(item.elementInstance);
      continue;
    }

    const contextElement = compositeElementWithContext(
      id,
      item.elementFn,
      item.appxContext,
      item.pendingProps,
    );
    children.push(contextElement);
    item.elementInstance = contextElement;
    item.pendingProps = null;
  }

  const parent = React.createElement('div', {}, children);
  return parent;
}

export function functionalMiniElement<TProps>(
  element: TElementFunction<TProps>,
  displayName = '' /* 用于问题排查，和小程序 axml 无关 */,
  elementType: EElementType,
  defaultProps?: TProps,
  _targetPlatform?: ETargetPlatform,
) {
  let targetPlatform: ETargetPlatform =
    _targetPlatform ?? ETargetPlatform.alipay;

  // 配置客户端环境
  const {
    pageEvents,
    componentEvents,
    buildOptions,
    componentLifeCycleToMount,
    componentLifeCycleToUnmount,
    pageLifeCycleToMount,
    pageLifeCycleToUnmount,
    getPropsFromInstance,
  } = platformConfig[targetPlatform];

  displayName = displayName || element.name;
  if (!displayName) {
    console.warn(
      '为了方便问题排查，请传入组件的 displayName 参数，或不要使用匿名函数组件 https://medium.com/@stevemao/do-not-use-anonymous-functions-to-construct-react-functional-components-c5408ec8f4c7',
    );
  }

  const nameTag = `[${elementType}/${displayName || '(unnamed)'}]`;
  const log = (...args) => {
    globalLog(nameTag, ...args);
  };
  const logErrorAndThrow = (err: Error) => {
    if (!err) return;
    err.message = `${nameTag} ${err.message}`;
    globalError(err);
    throw err;
  };

  if (elementType === 'component') {
    if (getComponent2Status() === EComponent2Status.UNKNOWN) {
      updateComponent2Status();
    }
    if (getComponent2Status() === EComponent2Status.INVALID) {
      // 如果还是 Unknown，就先不管了
      throw new Error(
        `无法注册 ${nameTag} 组件，因为当前小程序环境未开启 component2。配置入口在 mini.project.json : { "compileOptions": { "component2": true } }，或在 IDE 详情>项目配置中勾选`,
      );
    }
  }

  const platformExposedEvents: string[] =
    elementType === 'page' ? pageEvents : componentEvents;

  const defaultPropKeys = Object.keys(defaultProps || {});
  let observers = {};
  const elementMap: IInstanceMap = {};
  let commonTestRenderer;
  function updateReactTree() {
    act(() => {
      const parent = flushReactTree(elementMap);
      if (!commonTestRenderer) {
        commonTestRenderer = mountElement(parent);
      } else {
        commonTestRenderer.update(parent);
      }
    });
  }

  const WrappedElementFn = function (props: any) {
    let miniData: any;
    if (props && props[DANGER_ZONE_BYPASS_FUNCTION_CALL_WITH_DATA]) {
      miniData = props[DANGER_ZONE_BYPASS_FUNCTION_CALL_WITH_DATA];
      if (typeof miniData === 'string' && miniData.indexOf('%7B') === 0) {
        miniData = JSON.parse(decodeURIComponent(miniData));
      }
    } else {
      try {
        miniData = element.call(undefined, props);
      } catch (e) {
        e.message = `渲染出错 ${e.message}`;
        return logErrorAndThrow(e);
      }
    }
    if (typeof miniData === 'undefined') {
      log('函数组件没有返回渲染数据，请检查代码逻辑');
      miniData = {};
    } else if (typeof miniData !== 'object') {
      const e = new Error(
        `函数组件返回的渲染数据不合法，收到的类型为 ${typeof miniData}`,
      );
      return logErrorAndThrow(e);
    }
    // 不允许和 props 里的 key 重复
    const conflictKeys: string[] = [];
    const dataKeys = Object.keys(miniData) || [];
    for (const key in props) {
      if (dataKeys.indexOf(key) >= 0) {
        conflictKeys.push(key);
      }
    }
    if (conflictKeys.length > 0) {
      const e = new Error(
        `小程序自定义组件返回的渲染数据和 props 里的 key 重复：${conflictKeys.join(
          ', ',
        )}`,
      );
      return logErrorAndThrow(e);
    }

    // 上面的 early-return 都是检查和抛错，忽略 hooks 规则
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useSyncMiniData(miniData || {});
    return React.createElement('div', {}, FUNCTIONAL_MINI_PAGE_DOM_PLACEHOLDER);
  };

  // 在 onload 的时候，正式创建一个 react 组件
  function hookLoadToMountReactComponent(this: any, ...args: any[]) {
    const appxInstance = this;
    const id = getIdFromAppxInstance(appxInstance);
    if (elementMap[id])
      throw new Error(
        `duplicate id of appx instance, this might be a bug of minifish hooks. id: ${id}`,
      );
    const context = generateInstanceContext(appxInstance, false);
    appxInstance[instanceKeyPropNames] = defaultPropKeys;
    log('will mount react component');
    let initProps = {};
    // 微信和支付宝的行为是有所不同的：微信这时候还是 defaultProps 的内容，支付宝已经是真的数据了
    if (elementType === 'component') {
      initProps = getPropsFromInstance(appxInstance, defaultPropKeys);
    }

    elementMap[id] = {
      appxContext: context,
      appxId: id,
      elementFn: WrappedElementFn,
      pendingProps: initProps,
      unmounted: false,
      propKeys: defaultPropKeys,
    };
    updateReactTree();

    // 触发 onLoad
    if (elementType === 'page') {
      log('React 页面（Page）已经 Mount，开始触发 onLoad');
      return context.handlersController.callHandlers(
        pageLifeCycleToMount,
        appxInstance,
        args,
      ); // 直接调用controller 内部的方法，插队执行 onLoad
    } else {
      log(
        `React 组件（Component）已经 Mount，开始触发 ${componentLifeCycleToMount}`,
      );
      return context.handlersController.callHandlers(
        componentLifeCycleToMount,
        appxInstance,
        args,
      );
    }
  }

  function dispatchNewProps(appxInstance, nextProps) {
    const id = getIdFromAppxInstance(appxInstance);
    const instance = elementMap[id];
    instance.pendingProps = nextProps;
    updateReactTree();
  }

  function hookUnloadToUnmount(this: any) {
    const appxInstance = this;
    const id = getIdFromAppxInstance(appxInstance);
    log(`will unmount react element of ${id}`);
    if (!elementMap[id]) {
      log(
        `找不到 id 为 ${id} 的实例，请检查是否开启了 component2 。这也可能是 Minifish 的 bug。`,
      );
    }
    elementMap[id].unmounted = true;
    updateReactTree();
    // TODO: 清理 appx context
  }

  const handlersController = new HandlersController(nameTag);
  const anyUnknownContext = (ctx) => {
    if (!ctx) throw new Error('ctx is required');
    const id = getIdFromAppxInstance(ctx);
    return !elementMap[id];
  };
  const anyContext = () => true;
  if (elementType === 'page') {
    handlersController.addHandler(
      pageLifeCycleToMount,
      anyUnknownContext,
      hookLoadToMountReactComponent,
    );
    handlersController.addHandler(
      pageLifeCycleToUnmount,
      anyContext,
      hookUnloadToUnmount,
    );
  } else {
    // 组件的关键生命周期
    handlersController.addHandler(
      componentLifeCycleToMount,
      anyUnknownContext,
      hookLoadToMountReactComponent,
    );
    handlersController.addHandler(
      componentLifeCycleToUnmount,
      anyContext,
      hookUnloadToUnmount,
    );
    // 把 props 变更分发到对应的组件里
    if (targetPlatform === ETargetPlatform.alipay) {
      handlersController.addHandler(
        'deriveDataFromProps',
        anyContext,
        function hookDeriveDataFromProps(this: any, nextProps) {
          if (this.props && this.props === nextProps) return; // 可能是setData触发的，不要死循环了
          return dispatchNewProps(this, nextProps);
        },
      );
    } else if (targetPlatform === ETargetPlatform.wechat) {
      const props = defaultPropKeys.join(', ');
      log(`将注册以下 key 的 props 更新：${props}`);
      observers = {
        // 忽略函数参数，直接从 this 里面找
        // 搞不懂小程序的设计逻辑，为什么非要在 props 里夹着 data
        [props](this: any, ...args) {
          log('observer is being called', args);
          const newProps = getPropsFromInstance(this, defaultPropKeys);
          return dispatchNewProps(this, newProps);
          //
        },
        ...observers,
        // []
      };
    }
  }

  // 做一次预渲染，获取所有 appx 需要的属性
  const generateInstanceContext = (
    instance,
    ifServerRender: boolean,
  ): IElementContext => {
    return {
      instance,
      handlersController,
      ifServerRender,
      debugLog: log,
      platformConfig: platformConfig[targetPlatform],
    };
  };

  // 收集 initData
  let initData = {};
  const fakeAppxInstance = {
    $id: `_minifish_hooks_pre_render_${Math.random()}`,
    setData(data) {
      initData = {
        ...(initData || {}),
        ...data,
      };
    },
  };
  const fakeCtx = generateInstanceContext(fakeAppxInstance, true);
  const serverEl = compositeElementWithContext(
    fakeAppxInstance.$id,
    WrappedElementFn,
    fakeCtx,
    defaultProps || {},
  );
  serverRender(serverEl);
  log('serverRendered with initData', initData);

  // 拼装对象，喂给 appx。产出的配置应该只能喂一次，喂多了 appx 不认。
  handlersController.lockHandlerNames();

  // 把生命周期和一般事件处理分开归类，因为组件的事件处理要多裹一层
  const lifeCycleHandlers = {};
  const userEventHandlers = {};
  const handlers = handlersController.getHandlersImplProxy();
  for (const name in handlers) {
    if (platformExposedEvents.indexOf(name) >= 0) {
      lifeCycleHandlers[name] = handlers[name];
    } else {
      userEventHandlers[name] = handlers[name];
    }
  }

  const finalOptions = buildOptions(
    elementType,
    defaultProps,
    initData,
    lifeCycleHandlers,
    userEventHandlers,
    null,
    observers,
  );
  log('element options', finalOptions);
  return finalOptions;
}

export function tellIfFunctionalMiniContent(elementJson) {
  return elementJson?.children[0] === FUNCTIONAL_MINI_PAGE_DOM_PLACEHOLDER;
}
