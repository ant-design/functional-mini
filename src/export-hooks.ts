import { React } from './r.js';
import { useEvent } from './hooks.js';

const { useState, useEffect, useMemo, useCallback } = React;

export { useCallback, useEffect, useMemo, useState };

export { useEvent };
