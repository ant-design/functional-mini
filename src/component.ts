import { functionalMiniElement, TElementFunction } from './element.js';
import { getLifeCycleHooks } from './hooks.js';
import { alipayComponentEvents, commonComponentEvents } from './platform.js';
import { EElementType, ETargetPlatform } from './types.js';

export function alipayComponent<TProps extends Record<string, any>>(
  element: TElementFunction<TProps>,
  defaultProps?: TProps,
) {
  return functionalMiniElement(
    element,
    '',
    EElementType.component,
    defaultProps,
    ETargetPlatform.alipay,
  );
}

export function wechatComponent<TProps extends Record<string, any>>(
  element: TElementFunction<TProps>,
  defaultProps?: TProps,
) {
  return functionalMiniElement(
    element,
    '',
    EElementType.component,
    defaultProps,
    ETargetPlatform.wechat,
  );
}

export const useAttached = getLifeCycleHooks(commonComponentEvents.attached);
export const useReady = getLifeCycleHooks(commonComponentEvents.ready);
export const useMoved = getLifeCycleHooks(commonComponentEvents.moved);
export const useDetached = getLifeCycleHooks(commonComponentEvents.detached);

export const useDidMount = getLifeCycleHooks(
  alipayComponentEvents.didMount,
  undefined,
  ETargetPlatform.alipay,
);

// onInit, deriveDataFromProps, didUnmount, didUpdate 是不开放 Hooks 的，请使用 React 生命周期方法

export * from './export-hooks.js';

export { useComponent } from './hooks.js';
