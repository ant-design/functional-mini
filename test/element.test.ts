/*
基本的功能测试，组件相关的都靠 Page 和 Component 来保障
*/
import { afterEach, describe, expect, test, vi } from 'vitest';
import { IInstanceMap, flushReactTree } from '../src/element';
import { IElementContext } from '../src/hooks';
import {
  useEffect,
  useState,
  act,
  mountElement,
  createElement,
} from '../src/r';

import { delay } from './utils';

describe('element', () => {
  afterEach(() => {
    vi.clearAllTimers();
  });

  test('test unmount', async () => {
    const off = vi.fn();
    const Component = function (props: { foo: string }) {
      const [, setCounter] = useState(0);

      useEffect(() => {
        setCounter(2);
        setTimeout(() => {
          setCounter(10);
        }, 10);
        return off;
      }, []);

      return createElement('span', {}, 'Title');
    };

    let fakeRoot;
    act(() => {
      const reactEl = createElement(Component, { foo: 'test' });
      fakeRoot = mountElement(reactEl);
    });
    const rootString = fakeRoot.toString();
    expect(rootString).toMatchSnapshot();

    act(() => {
      fakeRoot.unmount();
    });

    expect(off).toHaveBeenCalledTimes(1);
  });

  test('basic render content', async () => {
    const C = function (props: { name: string }) {
      const [counterToShow] = useState(0);

      return createElement('div', {}, counterToShow);
    };

    let fakeRoot;
    act(() => {
      const reactEl = createElement(C, { name: 'test' });
      fakeRoot = mountElement(reactEl);
    });

    expect(fakeRoot.toString()).toMatchSnapshot();
  });

  test('test async action with .act', async () => {
    let internalSet;
    const C = function (props: { name: string }) {
      const [counter, setCounter] = useState(0);
      const [counterToShow, setCounterToShow] = useState('');

      useEffect(() => {
        internalSet = setCounter;
      }, []);

      useEffect(() => {
        setCounterToShow(`show-${counter}`);
      }, [counter]);

      return createElement('div', {}, counterToShow);
    };

    let fakeRoot;
    act(() => {
      const reactEl = createElement(C, { name: 'test' });
      fakeRoot = mountElement(reactEl);
    });
    setTimeout(() => {
      internalSet(2);
    }, 1);

    await delay(20);

    expect(fakeRoot.toString()).toMatchSnapshot();
  });

  test('basic react render', async () => {
    const initName = 'initName';
    const newName = 'newName';
    const C = function (props: { name: string }) {
      const [counter, setCounter] = useState(0);

      useEffect(() => {
        setCounter(counter + 1);
      }, [props?.name]);

      return createElement('div', {}, counter);
    };

    // 同一个组件，更新 props
    let fakeRoot;
    act(() => {
      const reactEl = createElement(C, {
        name: initName,
      });
      fakeRoot = mountElement(reactEl);
    });
    expect(fakeRoot.toString()).toMatchSnapshot();

    await delay(50);

    act(() => {
      const reactEl = createElement(C, {
        name: newName,
      });
      fakeRoot.update(reactEl);
    });
    expect(fakeRoot.toString()).toMatchSnapshot();

    // 用新的 root ，一定是新对象
    let newRoot;
    act(() => {
      const reactEl = createElement(C, {
        name: newName,
      });
      newRoot = mountElement(reactEl);
    });
    expect(newRoot.toString()).toMatchSnapshot();
  });

  test('component 交叉更新，会出错', async () => {
    // 父页面
    let parentSetState;
    let outerFn;
    const Parent = function () {
      const [s, setS] = useState('init');
      useEffect(() => {
        parentSetState = setS;
      }, [setS]);

      useEffect(() => {
        outerFn && outerFn();
      }, [s]);
      return createElement('div', {}, s);
    };

    // .create 需要在 .act 里面触发
    act(() => {
      const parentEl = createElement(Parent, {});
      mountElement(parentEl);
    });

    // 内部组件
    let childInternalSetName;
    const C = function (props) {
      const [name, setName] = useState('init');
      const [childName, setChildName] = useState('init');
      useEffect(() => {
        childInternalSetName = setChildName;
      });
      useEffect(() => {
        setName(props.query || 'empty');
      }, [props.query]);

      return createElement('div', {}, name + props.foo + childName);
    };

    let childRoot;
    act(() => {
      const reactEl = createElement(C, {
        foo: '1',
      });
      childRoot = mountElement(reactEl);
    });

    childInternalSetName('newName1');
    await delay(5);
    expect(childRoot.toString()).toMatchSnapshot();

    outerFn = () => {
      console.log('calling childInternalSetName');
      childInternalSetName('newName2');
    };

    parentSetState('2');
    await delay(20);
    expect(childRoot.toString()).toMatchSnapshot();
  });

  // 暴力刷一下，防止概率性失败
  let counter = 20;
  while (counter-- > 0) {
    test(`nested component ${counter}`, async () => {
      // 父页面
      let internalSet;
      let outerFn;
      const Parent = function () {
        const [s, setS] = useState('init');
        internalSet = setS;

        useEffect(() => {
          outerFn && outerFn();
        }, [s]);
        return createElement('div', {}, s);
      };

      // .create 需要在 .act 里面触发
      act(() => {
        const parentEl = createElement(Parent, {});
        mountElement(parentEl);
      });

      // 内部组件
      let childInternalSetName;
      const C = function (props) {
        const [name, setName] = useState('init');
        const [childName, setChildName] = useState('init');
        childInternalSetName = setChildName;

        useEffect(() => {
          setName(props.query || 'empty');
        }, [props.query]);

        return createElement('div', {}, name + props.foo + childName);
      };

      let childRoot;
      act(() => {
        const reactEl = createElement(C, {
          foo: '1',
        });
        childRoot = mountElement(reactEl);
      });

      childInternalSetName('newName1');
      await delay(5);
      expect(childRoot.toString()).toMatchSnapshot();

      outerFn = () => {
        childInternalSetName('newName2');
      };

      internalSet('2');
      await delay(50); // 这里如果写10，有可能会失败

      expect(childRoot.toString()).toEqual('<div>empty1newName2</div>');
    });
  }

  test('basic react update', async () => {
    const fn = vi.fn();
    const off = vi.fn();
    const Element = () => {
      useEffect(() => {
        fn();
        return off;
      });
      return createElement('div', {});
    };

    const props = { key: 'a' };
    let testRenderer;
    act(() => {
      const el = createElement(Element, props);
      testRenderer = mountElement(el);
    });

    expect(fn).toHaveBeenCalledTimes(1);
    expect(off).toHaveBeenCalledTimes(0);

    act(() => {
      // 多创建一次就是多跑一次函数，除非复用 element
      const el = createElement(Element, props);
      testRenderer.update(el);
    });

    expect(fn).toHaveBeenCalledTimes(2);
    expect(fn).toHaveBeenCalledTimes(2);
    expect(off).toHaveBeenCalledTimes(1);
  });

  test('component 都挂在一棵树上', async () => {
    // 父页面
    let internalSet;
    let outerFn;
    const A = function () {
      const [s, setS] = useState('init');
      useEffect(() => {
        internalSet = setS;
      }, []);

      useEffect(() => {
        outerFn && outerFn();
      }, [s]);
      return createElement('div', {}, s);
    };

    // 内部组件
    let childInternalSetName;
    const B = function (props) {
      const [name, setName] = useState('init');
      const [childName, setChildName] = useState('init');
      useEffect(() => {
        childInternalSetName = setChildName;
      });
      useEffect(() => {
        setName(props.query || 'empty');
      }, [props.query]);

      return createElement('div', {}, name + props.foo + childName);
    };

    let root;
    act(() => {
      const a = createElement(A, { key: 'a' });
      const b = createElement(B, { foo: 1, key: 'b' });
      const reactEl = createElement('div', {}, [a, b]);
      root = mountElement(reactEl);
    });

    childInternalSetName('newNameAA');
    await delay(5);
    expect(root.toString()).toMatchSnapshot();

    outerFn = () => {
      childInternalSetName('newName2');
    };

    internalSet('2');
    await delay(5);
    expect(root.toString()).toMatchSnapshot('empty1newName2');
  });

  test('basic element compose', async () => {
    const mockContext: IElementContext = {
      instance: {
        $id: 'aa',
        hanldersControllerts: null,
      },
    };

    const fn = vi.fn();
    const E = () => {
      useEffect(() => {
        fn();
      });
    };
    const elementMap: IInstanceMap = {
      aa: {
        appxContext: mockContext,
        appxId: 'aa',
        elementFn: E,
        pendingProps: {},
        unmounted: false,
      },
    };

    let testRenderer;
    act(() => {
      const parent = flushReactTree(elementMap);
      testRenderer = mountElement(parent);
    });
    await delay(5);
    expect(fn).toHaveBeenCalledTimes(1);

    // 应该不会重复渲染
    act(() => {
      const parent = flushReactTree(elementMap);
      testRenderer.update(parent);
    });
    await delay(5);
    expect(fn).toHaveBeenCalledTimes(1);
  });

  test('catch render error', async () => {
    const C = function () {
      throw new Error('test error');
    };

    expect(() => {
      mountElement(createElement(C, { shouldThrow: false }));
    }).toThrow();
  });

  test('mount element and act error', async () => {
    const C = function (props?: { shouldThrow: boolean }) {
      useEffect(() => {
        if (props?.shouldThrow) {
          throw new Error('test error');
        }
      });
      useEffect(() => {}, []);
      return createElement('div', {}, 'hello');
    };

    expect(() => {
      const instance = mountElement(createElement(C, { shouldThrow: false }));
      let newInstance;
      act(() => {
        newInstance = createElement(C, { shouldThrow: true });
        instance.update(newInstance);
      });
    }).toThrow();
  });
});
