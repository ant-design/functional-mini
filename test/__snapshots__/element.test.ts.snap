// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`element > basic react render 1`] = `"<div>1</div>"`;

exports[`element > basic react render 2`] = `"<div>2</div>"`;

exports[`element > basic react render 3`] = `"<div>1</div>"`;

exports[`element > basic render content 1`] = `"<div>0</div>"`;

exports[`element > component 交叉更新，会出错 1`] = `"<div>empty1newName1</div>"`;

exports[`element > component 交叉更新，会出错 2`] = `"<div>empty1newName2</div>"`;

exports[`element > component 都挂在一棵树上 > empty1newName2 1`] = `"<div><div>2</div><div>empty1newName2</div></div>"`;

exports[`element > component 都挂在一棵树上 1`] = `"<div><div>init</div><div>empty1newNameAA</div></div>"`;

exports[`element > nested component 0 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 1 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 2 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 3 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 4 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 5 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 6 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 7 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 8 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 9 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 10 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 11 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 12 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 13 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 14 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 15 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 16 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 17 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 18 1`] = `"<div>empty1newName1</div>"`;

exports[`element > nested component 19 1`] = `"<div>empty1newName1</div>"`;

exports[`element > test async action with .act 1`] = `"<div>show-2</div>"`;

exports[`element > test unmount 1`] = `"<span>Title</span>"`;
